File: bindings/csharp/src/TonAmount.cs
//
// The one value type the generated models do not give us.
//
// walletkit represents amounts as decimal strings (TokenAmount), which is right
// on the wire and wrong in an app: this keeps them exact.
//
using System.Globalization;
using System.Numerics;

namespace Ton.WalletKit
{
    /// <summary>
    /// An amount in nanotons. Kept as an integer rather than a decimal because TON
    /// amounts are exact integers and rounding one is a money bug.
    /// </summary>
    public readonly struct TonAmount
    {
        /// <summary>The raw value, in nanotons (1 TON = 1e9 nanotons).</summary>
        public BigInteger Nanotons { get; }

        /// <summary>Wraps a raw nanoton value.</summary>
        public TonAmount(BigInteger nanotons)
        {
            Nanotons = nanotons;
        }

        /// <summary>Parses a decimal nanoton string; zero when unparseable.</summary>
        public static TonAmount FromNanotons(string value)
        {
            return BigInteger.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var parsed)
                ? new TonAmount(parsed)
                : new TonAmount(BigInteger.Zero);
        }

        /// <summary>The nanoton value as the ABI expects it.</summary>
        public string ToRawString() => Nanotons.ToString(CultureInfo.InvariantCulture);

        /// <summary>Whole TON, for display. Never use this for arithmetic.</summary>
        public decimal ToTon() => (decimal)Nanotons / 1_000_000_000m;

        /// <inheritdoc/>
        public override string ToString() => ToTon().ToString("0.#########", CultureInfo.InvariantCulture) + " TON";
    }
}

File: bindings/android/src/main/java/org/ton/walletkit/core/TonAmount.java
package org.ton.walletkit.core;

import java.math.BigDecimal;
import java.math.BigInteger;

/**
 * An amount in nanotons.
 *
 * <p>An integer, not a double: TON amounts are exact and rounding one is a money
 * bug. walletkit sends them as decimal strings for the same reason.
 */
public final class TonAmount {

    private static final BigDecimal NANO = new BigDecimal("1000000000");

    private final BigInteger nanotons;

    public TonAmount(BigInteger nanotons) {
        this.nanotons = nanotons == null ? BigInteger.ZERO : nanotons;
    }

    /** Parses a decimal nanoton string; zero when unparseable. */
    public static TonAmount fromNanotons(String value) {
        try {
            return new TonAmount(new BigInteger(value));
        } catch (RuntimeException malformed) {
            return new TonAmount(BigInteger.ZERO);
        }
    }

    public BigInteger getNanotons() {
        return nanotons;
    }

    /** The value as the ABI expects it. */
    public String toRawString() {
        return nanotons.toString();
    }

    /** Whole TON, for display. Never use this for arithmetic. */
    public BigDecimal toTon() {
        return new BigDecimal(nanotons).divide(NANO);
    }

    @Override
    public String toString() {
        return toTon().stripTrailingZeros().toPlainString() + " TON";
    }
}
